### PR TITLE
feat(ica/controller)!: migrate ica/controller parameters to be self managed

### PR DESCRIPTION
**Pull request by vishal-kanna**
## Description

With this PR:

- [x] The `ica/controller` submodule now self-manages its parameters.
- [x] Unit and integration tests have been added for the `ica/controller` submodule's parameters.
- [x] Migration code has been implemented for moving parameters from `x/params`' subspace to the `ica/controller` submodule's keeper.
- [x] Migration unit and integration tests have been added.

closes: #3503  

addresses: #2010 

### Commit Message / Changelog Entry

```text
feat(ica/controller)!: migrate ica/controller parameters to be self managed
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by damiannolan** [comment]
Hey @vishal-kanna! Thanks for getting this started! There has been many comments on #3520 which could be useful to align the work in this PR. Let us know if you'd like to continue working on this, alternatively we could have someone pick up the remaining work and bring it over the finish line!

**Comment by srdtrk** [comment]
Thanks for the work @vishal-kanna. I will take over the remaining work now.

**Review by srdtrk — APPROVED** [review]

File: modules/apps/27-interchain-accounts/controller/keeper/keeper_test.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper_test.go:254]:
```suggestion
		{"success: set params true", types.NewParams(true), true},
```

File: modules/apps/27-interchain-accounts/controller/keeper/migrations.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/keeper/migrations.go:44]: nit: might be useful to keep `controllertypes` alias

File: modules/apps/27-interchain-accounts/controller/keeper/msg_server.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/keeper/msg_server.go:77]: should we make use of `k.GetAuthority()`?

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/controller/keeper/keeper.go:284]: I would still advocate for the following:

```suggestion
	if len(bz) == 0 {
		return types.Params{}
	}
```

This came up in the client params PR also - see https://github.com/cosmos/ibc-go/pull/3640#discussion_r1204210620

File: modules/apps/27-interchain-accounts/controller/keeper/migrations.go
Comment (DimitrisJim) [inline modules/apps/27-interchain-accounts/controller/keeper/migrations.go:54]:
```suggestion
// MigrateParams migrates the controller submodule's parameters from the x/params to self store.
```

File: modules/apps/27-interchain-accounts/module.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/module.go:157]: but golanci-lint complains if I return nil right after

File: proto/ibc/applications/interchain_accounts/controller/v1/tx.proto
Comment (srdtrk) [reply proto/ibc/applications/interchain_accounts/controller/v1/tx.proto:55]: I think it means the passed value from `app.go` is `x/gov` account by default

File: modules/apps/27-interchain-accounts/controller/keeper/migrations.go
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/controller/keeper/migrations.go:44]: generally yes, but iirc this came up a few times when we implemented ICA and due to there being the two submodules it can be useful to be explicit to differentiate. I don't feel particularly strongly about it tho

File: modules/apps/27-interchain-accounts/controller/keeper/keeper.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/keeper/keeper.go:284]: But this would be inconsistent with the host and transfer PRs. Should we at least preserve consistency within the ica module? XD

File: modules/apps/27-interchain-accounts/controller/keeper/migrations.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/keeper/migrations.go]: done

File: modules/apps/27-interchain-accounts/controller/ibc_middleware.go
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/ibc_middleware.go:7]: super nit: I would consider external modules to be any module in the go.mod import, but this is fine with me as well

**Review by colin-axner — COMMENTED** [review]
LGTM! Just one minor fix related to the controller or host keeper potentially being nil (it is allowed to have the ica module with just one of the keepers initialized)

File: modules/apps/27-interchain-accounts/genesis/types/genesis.go
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/genesis/types/genesis.go:80]: Why was the validation of params removed? For host params the validation is still there.

Comment (srdtrk) [reply modules/apps/27-interchain-accounts/genesis/types/genesis.go:80]: Validation is not needed for parameters that are just booleans. Because all booleans are valid. Validation was also removed from transfer. The validation was needed when we used `x/params` in case the external module did something funny. In host module, the params weren't just bools

File: modules/apps/27-interchain-accounts/module.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/module.go:157]: Isn't that checked above? In lines 136-145

Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/module.go:157]: I think @colin-axner means that `hostMigrator.MigrateParams` and `controllerMigrator.MigrateParams` both use the host and controller keeper respectively and, by the look of the code, both `hostkeeper.NewMigrator` and `controllerkeeper.NewMigrator` could potentially be passed a `nil` keeper...

Comment (damiannolan) [reply modules/apps/27-interchain-accounts/module.go:157]: Yeah, the code here code use a little restructuring imo for better readability/maintainability. 

The existing code was fine because the keeper is nil checked in the controller migration [here](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/migrations.go#L28) for `AssertChannelCapabilityMigrations`. 

It may be nicer to move the instantiation of the migrator types into the nil checks above at the top of this function.

Comment (srdtrk) [reply modules/apps/27-interchain-accounts/module.go:157]: Is the worry here that a keeper won't be initialized? Or incorrect wiring will take place? I mean, there's gotta be some sort of error if keeper is nil right? Sorry, I'm lost here. Or do we want keeper to be able to be nil?

Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/module.go:157]: Both the controller and the host keeper are allowed to be `nil` because chains might choose to to use one or the other submodule. See [this section of the documentation](https://ibc.cosmos.network/main/apps/interchain-accounts/integration.html#using-submodules-exclusively) for more info.

Comment (srdtrk) [reply modules/apps/27-interchain-accounts/module.go:157]: I've moved the check inside the migration functions

**Review by damiannolan — APPROVED** [review]
Awesome, thank you @vishal-kanna and @srdtrk :) 

One last thing would be logging msgs for both cases - successful migration and if keeper is nil and migrations are skipped. But I'm happy if they are handled as part of https://github.com/cosmos/ibc-go/issues/3776

**Comment by srdtrk** [comment]
I will merge this after your approval @DimitrisJim

File: modules/apps/27-interchain-accounts/controller/types/msgs.go
Comment (DimitrisJim) [inline modules/apps/27-interchain-accounts/controller/types/msgs.go:94]: do you remember if this caused an issue? In general, seems like interfacer has been [deprecated](https://github.com/mvdan/interfacer). Unsure if this is a relic from the past?

**Review by DimitrisJim — APPROVED** [review]
looks great :heart: left a small question for something unrelated.

File: modules/apps/27-interchain-accounts/controller/types/msgs.go
Comment (srdtrk) [reply modules/apps/27-interchain-accounts/controller/types/msgs.go]: done

**Pull request merged**
